### ISSUE TITLE
complete: support json for complex arguments

### ISSUE THREAD

**Issue by rsteube** [fish-shell/fish-shell#10197]
So in [carapace](https://carapace.sh) I've got data like [this](https://rsteube.github.io/carapace/carapace/export.html).
Would be nice if fish would provide a bit more control over the completion values.

What might work is accepting `json` additionally to the `value\tdescription` format.
Meaning: check if response starts with `{` and try to parse the result as `json` otherwise fallback to the default.

## Display

What to show during completion.
Fish currently truncates the value without knowing where to cut.

![image](https://github.com/fish-shell/fish-shell/assets/9090290/872da34d-52ae-42a3-8583-7e3850b1a42d)

![image](https://github.com/fish-shell/fish-shell/assets/9090290/b3c9d66c-7001-4891-b745-d5d6ca7e0ff3)

## Style

How to highlight the displayed value (color, bold, underlined...).
Either a separate value or support ansi esscape codes for display.

![image](https://github.com/fish-shell/fish-shell/assets/9090290/91138f95-42d9-4570-a001-96606fcb658c)

![image](https://github.com/fish-shell/fish-shell/assets/9090290/5832728e-da3b-44ef-b68f-9a0b830e5e5d)

## Space Suffix

Some way to control if a space suffix shall be added. E.g. when completing a command that is passed as string.

![image](https://github.com/fish-shell/fish-shell/assets/9090290/cf4b6b2f-e1c2-4e7d-9330-f45feb22fcd1)

![image](https://github.com/fish-shell/fish-shell/assets/9090290/3913ebc0-c482-47d5-a8c5-d81dfc509e13)

## Error / Usage

Some way to show error and usage messages separate from the completion values.
Errors are currently embedded into the values as workaround.

![image](https://github.com/fish-shell/fish-shell/assets/9090290/dcf046fd-75db-4a10-9fad-edc5e2615b0d)

![image](https://github.com/fish-shell/fish-shell/assets/9090290/ce172342-f8d0-4888-9b20-2a0d16ccc39b)

## Tags

Additional identifier for the values like `subcommand group`, `files`, `branches` to filter or group them.

![image](https://github.com/fish-shell/fish-shell/assets/9090290/2ca5c0f1-3383-43a1-93d8-20189e977639)

![image](https://github.com/fish-shell/fish-shell/assets/9090290/2e42415c-8dc1-4737-bb19-328705d7496f)

related #7832

**Comment by krobelus** [comment]
> Some way to control if a space suffix shall be added.

A longstanding issue is that serialized completions don't carry metadata.
I agree we should try to add them and propagate flags like `NO_SPACE`.

>  Either a separate value or support ansi esscape codes for display.

Once we have flags we can add one to allow to disable coloring for individual completions.

> Fish currently truncates the value without knowing where to cut.

I'd rather see a generalized improvement here (also we want to be somewhat honest about what will be inserted).
I tried changing this to only show what's not on the command line but it didn't look great in some cases.
The current behavior is a nice compromise but maybe we can improve it.

> Errors are currently embedded into the values as workaround.
> Additional identifier for the values like subcommand group, files, branches to filter or group them.

Hmm, we could add a text field (separate from the completion pager), that should work for both?

**Comment by rsteube** [comment]
> I'd rather see a generalized improvement here (also we want to be somewhat honest about what will be inserted). I tried changing this to only show what's not on the command line but it didn't look great in some cases. The current behavior is a nice compromise but maybe we can improve it.

True, but I don't think there is much room for improvement here without having additional context from the completion scripts.
To be clear: In my case I do have the data for this as there is a separation between displayed and inserted value (see [Prefix](https://rsteube.github.io/carapace/carapace/action/prefix.html) / [Suffix](https://rsteube.github.io/carapace/carapace/action/suffix.html)).

> Hmm, we could add a text field (separate from the completion pager), that should work for both?

Yes, that is how it is done it zsh: one multiline string with different coloring.

**Comment by mqudsi** [comment]
Note that displaying errors is possible (with color or what have you) but not "natively" via the completions machinery. See for example the `npm` completions:

```fish
function __npm_helper_installed
    # This function takes the command to globally install a package as $argv[1]
    if not type -q all-the-package-names
        if not set -qg __fish_npm_pkg_info_shown
            set -l old (commandline)
            commandline -r ""
            echo \nfish: Run `$argv[1] all-the-package-names` to gain intelligent \
                package completion >&2
            commandline -f repaint
            commandline -r $old
            set -g __fish_npm_pkg_info_shown 1
        end
        return 1
    end
end

```
